photosynthesis_metrics/msid.py

```diff
@@ -0,0 +1,420 @@
+              normalized_laplacian=True, normalize='empty'):
+    '''
+    Compute the msid descriptor for a single sample x
+    Arguments:
+        x: x samples
+        ts: temperature values
+        k: number of neighbours for graph construction
+        m: Lanczos steps in SLQ
+        niters: number of starting random vectors for SLQ
+        rademacher: if True, sample random vectors from Rademacher distributions, else sample standard normal distribution
+        graph_builder: if 'kgraph', uses faster graph construction (options: 'sparse', 'kgraph')
+        normalized_laplacian: if True, use normalized Laplacian
+        normalize: 'empty' for average heat kernel (corresponds to the empty graph normalization of NetLSD),
+                'complete' for the complete, 'er' for erdos-renyi
+                normalization, 'none' for no normalization
+    Returns:
+        normed_msidx: normalized msid descriptor
+    '''
+    Lx = _build_graph(x, k, graph_builder, normalized_laplacian)
+
+    nx = Lx.shape[0]
+    msidx = slq_red_var(Lx, m, niters, ts, rademacher)
+
+    normed_msidx = _normalize_msid(msidx, normalize, nx, k, ts) * NORMALIZATION
+
+    return normed_msidx
+
+class MSID(nn.Module):
+    r"""Creates a criterion that measures MSID score for a batch of images
+    See https://arxiv.org/abs/1905.11141 for reference.
+
+
+    Args:
+        ts: temperature values
+        k: number of neighbours for graph construction
+        m: Lanczos steps in SLQ
+        niters: number of starting random vectors for SLQ
+        rademacher: if True, sample random vectors from Rademacher distributions, else sample standard normal distribution
+        graph_builder: if 'kgraph', uses faster graph construction (options: 'sparse', 'kgraph')
+        normalized_laplacian: if True, use normalized Laplacian
+        normalize: 'empty' for average heat kernel (corresponds to the empty graph normalization of NetLSD),
+                'complete' for the complete, 'er' for erdos-renyi
+                normalization, 'none' for no normalization
+        msid_reduction: 'l2' to compute the l2 norm of the distance between `predictoon` and `target`;
+                'max' to find the maximum abosulute difference between two descriptors over temperature
+        channel_reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed. 
+
+    """
+
+    def __init__(self, ts: List[float] = np.logspace(-1, 1, 256), k: int = 5, m: int = 10, niters: int = 100, rademacher: bool = False, graph_builder: str = 'sparse',
+              normalized_laplacian: bool = True, normalize: str = 'empty', msid_reduction: str = "max", channel_reduction: str = 'mean'):
+
```

Comment:
Formatting: remove blank line here.

From the Google style guide (section 3.5):
Two blank lines between top-level definitions, be they function or class definitions. One blank line between method definitions and between the class line and the first method. **No blank line following a def line**. Use single blank lines as you judge appropriate within functions or methods.